### ISSUE TITLE
Ability to let apps do their own upgrades

### ISSUE THREAD

**Issue by aschrijver** [YunoHost/issues#2005]
Yunohost wants to make it as easy as possible to do sysadmin, and has great features to do so. It takes over the native application update / upgrade procedures via the Admin UI.

But some apps already have great and intuitive upgrade facilities and having Yunohost take that away is a step backward instead of an improvement. One such app are Discourse forums (see screenshots). Yunohost Discourse App has disabled the native update in the UI. If there's a problem with Discourse or cool new features available, now there's an extra dependency.. we have to wait until there's an updated Yunohost-App version to install.

Hence the feature request:

> **There should be the possibility to leave application updates to the application.**

Either app integrators could decide on case-by-case basis whether they want to use Yunohost or their own app, or maybe even end-users might toggle / choose the upgrade option at install time. Yuno would report "Updates handled by the app" in the UI if that was the case.

### Example: Discourse Native Upgrade UI

![Screenshot from 2022-04-17 09-04-45](https://user-images.githubusercontent.com/5111931/163704577-b5012f79-5c3a-43d5-ab50-fe02fd475233.png)

![Screenshot from 2022-04-17 09-05-38](https://user-images.githubusercontent.com/5111931/163704583-604e4c9a-da2c-417a-960c-85d3464acd1e.png)

---

See also:

- https://forum.yunohost.org/t/ability-to-skip-yunohost-update-process-and-leave-it-to-the-app/19531
- Issue I have that makes troubleshooting more complex, having to take Yunohost into account:
  - https://github.com/YunoHost-Apps/discourse_ynh/issues/108

**Comment by ericgaspar** [comment]
This doesn't seem possible for Discourse because we're patching the app on install (FYI, the patchs don't work anymore... that's why upgrading to 2.8.0 doesn't work right now).

**Comment by aschrijver** [comment]
Maybe it is not possible to retro-actively add or enable this feature. But one way I imagine this to work is that an app integrator (the developer) at the time of writing/coding 'the recipe' for their integration would consider like "Well, this app has a good upgrade experience, so I don't need Yunohost's override mechanism", and toggles something in config to disable it.

Or there may be a decision posed to the person installing the app, before the actual installation starts: "This app offers a good integrated upgrade experience. Do you want Yunohost to manage app upgrades instead? Yes/No". (Or vice versa: "Do you want Yunohost to delegate upgrades to the application? Yes/No").

**Comment by tituspijean** [comment]
With the current design of YunoHost, that would be up to the app packagers to decide what to do with upgrades.

For example, I let Radarr and [Sonarr](https://github.com/YunoHost-Apps/sonarr_ynh#disclaimers--important-information) (and maybe other apps in the *arr family) do their own thing with updates. However that is risky:

### Versions mismatch

Assuming we implement your proposal, we need a way to monitor the app actual version. Not all apps provide an easy way to get it. Some just tell you to fetch their git main branch and serve the files. So we would need to create custom hook scripts for almost each apps.

### Custom patches

@ericgaspar's counter-argument comes here. Moreover, assuming an app introduces breaking changes in one of its upgrades, we need to be able to perform whatever actions it requires us to do manually.

### Upgrades in continuous integration tests

> But some apps already have great and intuitive upgrade facilities

Until they don't. The whole argument here for sped up upgrades falls apart the moment a badly designed upgrade is released and corrupts user data. I am not saying our current way to do upgrades prevents this with 100% accuracy though. There are plenty of ways for upgrades to go badly. Letting us test the apps in most use and upgrade cases will prevent catastrophes.

---

In summary, our current way to control upgrades is slow, and may be frustrating for users getting a constant nag-for-upgrades from their apps. It however has it virtues, as in letting us test the upgrades and make sure they go smoothly. Overall we prefer to have "slightly" out-of-date apps than a flow of discontent users betrayed by a botched upgrade.

**Comment by alexAubin** [comment]
+1 to all titus's arguments ... basically we're trying to standardize and centralize app upgrades, just like your classic Linux distribution handles all the upgrade via the package manager ... The fact that some apps have their own self-upgrade mechanism is symptomatic of the chaotic self-hosting ecosystem were app devs want to handle "everything", including for some of them SSL certs ... To me this is reminiscent of the Windows chaos where there's no centralized upgrade system, hence every software has its own self-upgrade mechanism harassing you and slowly creating a bloated mess

**Comment by aschrijver** [comment]
I don't know about 'bloated mess' and was mostly talking about leaving the choice to the Yunohost app maintainer. In the case of Discourse there's a super well established, quality upgrade process, integrated in Discourse admin (which any Yunohost installer must be familiar with anyway, if they'll maintain the forum). This upgrade process offers not only new desirable features, but also critical security patches on regular basis. Yunohost on the premise of 'too hard' and 'bloated mess' disables an intuitive integrated update feature, and replaces it with one where there's a dependency on a number of volunteers (though absolutely kudos to them for building the integration) to keep things similarly well updated and stable. This is adding a point-of-failure in a chain, imho, and should be serious consideration to anyone that wants to host a Discourse forum.

**Comment by alexAubin** [comment]
The fact that Discourse has its own super-duper great self-upgrade workflow doesn't change the fact, as Titus explained, you risk breaking things such as LDAP/SSO integration, or ending up with an unpatched software that would hence not work anymore as expected if not break - and in some cases too, *security issues*. 

Surely if somebody was to upgrade app FooBar using its own self-upgrade mechanism and this broke whatever stuff, we would get admins coming on this very same bugtracker complaining that "omg if using the self-upgrade mechanism breaks my app, why are your not disabling the self-upgrade mechanism in the first place ?!"

When there are critical security patches, the Yunohost packager team tend to release new app versions in a timely manner.

Ultimately there's no magic solution to this : distributing software is an underestimated complex matter, and it requires coordination between distribution maintainers *AND* upsteam devs. But we're in a situation where upstream devs won't really care about stuff like YunoHost, because the defacto standard is "most of people are sysadmin running Docker and they'll know how to upgrade stuff manually, right?" - while at the same time there are many benefit to use a tool like YunoHost which integrates the app with SSO/LDAP (for UX), systemd conf hardening, Fail2ban (security), etc. And in the meantime our packaging format is not super well suited to be maintained by other upstream devs so.

**Comment by aschrijver** [comment]
Thanks for the response. By no means I mean to be pushy in having this, and esp. given there's no appetite to have this. I love what y'all are doing with Yunohost, and totally respect choices on overall project direction. I only will react to:

> "omg if using the self-upgrade mechanism breaks my app, why are your not disabling the self-upgrade mechanism in the first place ?!"

What I was referring to was an opt-in choice to override the Yunohost update process on a per-app basis to fall back to the apps' one. There might be a big red text warning "Are you sure? .. this can have dire consequences".
